# Data Mining Final Exam Part 2

    Name: Aimi Hassin

### Question 4

The objective of this question is to find frequent itemsets using the dataset from Question 1. In the context of my dataset, I would like to know which house_type that airbnb guests frequently book when they are in Manhattan and/or Orlando. Knowing which house_type has highest booking frequency may be helpful for the host to understand about the behaviour of airbnb users. Besides, it is also may be beneficial for the airbnb users too as they able to know what kinds of house_type dominate a particular place.

#### Import Required Libraries

In [390]:
import pandas as pd
import numpy as np
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

#### Load Data

In [391]:
df = pd.read_csv('airbnb.csv')
df.sample(5)

Unnamed: 0  listing_id  host_id       city  \
71           72        1071    99971  Manhattan   
242         249        1242   100142  Manhattan   
137         141        1137   100037  Manhattan   
330         338        1330   100230    Orlando   
540         552        1540   100440    Orlando   

                                           house_title  \
71      BEAUTIFUL 1 BEDROOM Heart of Upper East Side!!   
242  Quiet, tranquil room in the middle of everything!   
137                         Soho - Beautiful 3 Bedroom   
330   INTERNATIONAL dr/UNIVERSAL 2 Bedroom CONDO-HOTEL   
540        Cozy Condo by SeaWorld 1 Bedroom/1 Bath-OBD   

                    house_type  price owner_name  no_of_reviews  rating  \
71            Entire apartment   70.0     Albina             21     4.5   
242  Private room in apartment  149.0      Eddie            121     4.5   
137           Entire apartment  266.0    Kathrin            220     4.5   
330         Entire condominium   49.0    Gabriel             64     4.5   
540         Entire condominium   59.0       Parc             31     4.5   

                                             occupancy  no_of_guest  \
71   ['4 guests', '1 bedroom', '2 beds', '1 bath', ...          4.0   
242  ['4 guests', '1 bedroom', '1 bed', '2 shared b...          4.0   
137  ['5 guests', '3 bedrooms', '3 beds', '2 baths'...          5.0   
330  ['6 guests', '2 bedrooms', '4 beds', '2 baths'...          6.0   
540  ['4 guests', '1 bedroom', '2 beds', '1 bath', ...          4.0   

     bedrooms_no  beds_no  bath_no  
71           1.0      2.0      1.0  
242          1.0      1.0      2.0  
137          3.0      3.0      2.0  
330          2.0      4.0      2.0  
540          1.0      2.0      1.0

In [392]:
# Removing spaces from the 'house_type'
df['house_type'] = df['house_type'].str.strip()
df.sample(3)

Unnamed: 0  listing_id  host_id       city  \
275         282        1275   100175  Manhattan   
524         536        1524   100424    Orlando   
48           48        1048    99948  Manhattan   

                                      house_title                 house_type  \
275              NYC Bedroom in Amazing Location!  Private room in apartment   
524  Newly built, modern and cozy Suite near UCF.           Entire apartment   
48                 Sunny Private Studio-Apartment           Entire apartment   

     price     owner_name  no_of_reviews  rating  \
275  133.0       Sterling            143     5.0   
524   64.0  Camilo & Lina             89     5.0   
48   119.0      Alexandra            129     5.0   

                                             occupancy  no_of_guest  \
275  ['4 guests', '1 bedroom', '3 beds', '1 shared ...          4.0   
524  ['4 guests', '1 bedroom', '1 bed', '1 bath', '...          4.0   
48   ['4 guests', '1 bedroom', '2 beds', '1 bath', ...          4.0   

     bedrooms_no  beds_no  bath_no  
275          1.0      3.0      1.0  
524          1.0      1.0      1.0  
48           1.0      2.0      1.0

#### Creating A List

In [432]:
df1 = df[['house_type', 'city']]
df1

house_type       city
0             Entire apartment  Manhattan
1    Private room in apartment  Manhattan
2             Entire apartment  Manhattan
3    Private room in apartment  Manhattan
4             Entire apartment  Manhattan
..                         ...        ...
593           Entire townhouse    Orlando
594  Entire serviced apartment    Orlando
595         Entire guest suite    Orlando
596           Entire apartment    Orlando
597         Entire condominium    Orlando

[598 rows x 2 columns]

In [394]:
house_list = df1.values.tolist()
print(house_list)

[['Entire apartment', 'Manhattan'], ['Private room in apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Private room in apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Private room in apartment', 'Manhattan'], ['Private room in loft', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Private room in apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Private room in townhouse', 'Manhattan'], ['Private room in apartment', 'Manhattan'], ['Entire apartment', 'Manhattan'], ['Private room in townhouse', '

#### Converting List to DataFrame with Boolean Values

In [433]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_array = te.fit(house_list).transform(house_list)
df2 = pd.DataFrame(te_array, columns=te.columns_)

In [434]:
df2

Camper/RV  Entire apartment  Entire condominium  Entire guest suite  \
0        False              True               False               False   
1        False             False               False               False   
2        False              True               False               False   
3        False             False               False               False   
4        False              True               False               False   
..         ...               ...                 ...                 ...   
593      False             False               False               False   
594      False             False               False               False   
595      False             False               False                True   
596      False              True               False               False   
597      False             False                True               False   

     Entire guesthouse  Entire house  Entire loft  Entire serviced apartment  \
0                False         False        False                      False   
1                False         False        False                      False   
2                False         False        False                      False   
3                False         False        False                      False   
4                False         False        False                      False   
..                 ...           ...          ...                        ...   
593              False         False        False                      False   
594              False         False        False                       True   
595              False         False        False                      False   
596              False         False        False                      False   
597              False         False        False                      False   

     Entire townhouse  Entire villa  ...  Private room in cottage  \
0               False         False  ...                    False   
1               False         False  ...                    False   
2               False         False  ...                    False   
3               False         False  ...                    False   
4               False         False  ...                    False   
..                ...           ...  ...                      ...   
593              True         False  ...                    False   
594             False         False  ...                    False   
595             False         False  ...                    False   
596             False         False  ...                    False   
597             False         False  ...                    False   

     Private room in guest suite  Private room in guesthouse  \
0                          False                       False   
1                          False                       False   
2                          False                       False   
3                          False                       False   
4                          False                       False   
..                           ...                         ...   
593                        False                       False   
594                        False                       False   
595                        False                       False   
596                        False                       False   
597                        False                       False   

     Private room in house  Private room in loft  Private room in resort  \
0                    False                 False                   False   
1                    False                 False                   False   
2                    False                 False                   False   
3                    False                 False                   False   
4                    False                 False                   False   
..                     ...                   ...                     ...   
5

#### Find frequently occurring itemsets using Apriori Algorithm

In [420]:
from mlxtend.frequent_patterns import apriori

frequent_itemsets_ap = apriori(df2, min_support=0.05, use_colnames=True)

In [421]:
frequent_itemsets_ap

support                                itemsets
0   0.496656                      (Entire apartment)
1   0.102007                    (Entire condominium)
2   0.058528                          (Entire house)
3   0.498328                             (Manhattan)
4   0.501672                               (Orlando)
5   0.102007             (Private room in apartment)
6   0.065217                 (Private room in house)
7   0.349498           (Manhattan, Entire apartment)
8   0.147157             (Orlando, Entire apartment)
9   0.091973           (Orlando, Entire condominium)
10  0.056856                 (Orlando, Entire house)
11  0.093645  (Private room in apartment, Manhattan)
12  0.058528        (Orlando, Private room in house)

#### Find frequently occurring itemsets using F-P Growth

In [424]:
from mlxtend.frequent_patterns import fpgrowth

frequent_itemsets_fp=fpgrowth(df2, min_support=0.05, use_colnames=True)

In [425]:
frequent_itemsets_fp

support                                itemsets
0   0.498328                             (Manhattan)
1   0.496656                      (Entire apartment)
2   0.102007             (Private room in apartment)
3   0.065217                 (Private room in house)
4   0.102007                    (Entire condominium)
5   0.058528                          (Entire house)
6   0.501672                               (Orlando)
7   0.349498           (Manhattan, Entire apartment)
8   0.147157             (Orlando, Entire apartment)
9   0.093645  (Private room in apartment, Manhattan)
10  0.058528        (Orlando, Private room in house)
11  0.091973           (Orlando, Entire condominium)
12  0.056856                 (Orlando, Entire house)

#### Mine the Association Rules

   1) **Support** = the popularity of item and can be calculated by finding the number of transactions containing a particular item divided by the total number of transactions.
    
   2) **Confidence** = the likelihood that an item B is also bought if item A is bought. It can be calculated by finding the number of transactions where A and B are bought together, divided by the total number of transactions where A is bought. Mathematically, it can be represented as: Confidence(A → B) = (Transactions containing both (A and B))/(Transactions containing A)
    
   3) **Lift** = Lift refers to the increase in the ratio of the sale of B when A is sold. Lift(A –> B) can be calculated by dividing Confidence(A -> B) divided by Support(B). Mathematically it can be represented as: Lift(A→B) = (Confidence (A→B))/(Support (B))

In [429]:
from mlxtend.frequent_patterns import association_rules

rules_ap = association_rules(frequent_itemsets_ap, metric="confidence", min_threshold=0.7)
rules_fp = association_rules(frequent_itemsets_fp, metric="confidence", min_threshold=0.7)

In [430]:
rules_ap

antecedents         consequents  antecedent support  \
0                  (Manhattan)  (Entire apartment)            0.498328   
1           (Entire apartment)         (Manhattan)            0.496656   
2         (Entire condominium)           (Orlando)            0.102007   
3               (Entire house)           (Orlando)            0.058528   
4  (Private room in apartment)         (Manhattan)            0.102007   
5      (Private room in house)           (Orlando)            0.065217   

   consequent support   support  confidence      lift  leverage  conviction  
0            0.496656  0.349498    0.701342  1.412130  0.102001    1.685356  
1            0.498328  0.349498    0.703704  1.412130  0.102001    1.693144  
2            0.501672  0.091973    0.901639  1.797268  0.040799    5.066332  
3            0.501672  0.056856    0.971429  1.936381  0.027494   17.441472  
4            0.498328  0.093645    0.918033  1.842227  0.042813    6.120401  
5            0.501672  0.058528    0.897436  1.788889  0.025811    4.858696

In [431]:
rules_fp

antecedents         consequents  antecedent support  \
0                  (Manhattan)  (Entire apartment)            0.498328   
1           (Entire apartment)         (Manhattan)            0.496656   
2  (Private room in apartment)         (Manhattan)            0.102007   
3      (Private room in house)           (Orlando)            0.065217   
4         (Entire condominium)           (Orlando)            0.102007   
5               (Entire house)           (Orlando)            0.058528   

   consequent support   support  confidence      lift  leverage  conviction  
0            0.496656  0.349498    0.701342  1.412130  0.102001    1.685356  
1            0.498328  0.349498    0.703704  1.412130  0.102001    1.693144  
2            0.498328  0.093645    0.918033  1.842227  0.042813    6.120401  
3            0.501672  0.058528    0.897436  1.788889  0.025811    4.858696  
4            0.501672  0.091973    0.901639  1.797268  0.040799    5.066332  
5            0.501672  0.056856    0.971429  1.936381  0.027494   17.441472

**Comment:** From the results above, it is known that airbnb users tend to book 'entire apartment' or 'private room in apartment' when they are in Manhattan with confidence of 0.70 and 0.92 respectively. Meanwhile, in Orlando, the guests prefer to book 'entire house', 'entire condominium' or 'private room in house' with confidence ranging from 0.89 to 0.97. Here, we can understand that most of the booked airbnb listings in Manhattan are apartment-typed, while in Orlando the type of house would be a house or condominium.